hydra-node/src/Hydra/Chain/ScriptRegistry.hs

```diff
@@ -93,18 +98,75 @@ publishHydraScripts networkId socketPath sk = do
   systemStart <- querySystemStart networkId socketPath QueryTip
   eraHistory <- queryEraHistory networkId socketPath QueryTip
   stakePools <- queryStakePools networkId socketPath QueryTip
-  utxo <- queryUTxOFor networkId socketPath QueryTip vk
-  txs <- buildScriptPublishingTxs pparams systemStart networkId eraHistory stakePools utxo sk
+  txs <- buildHydraScriptTxs pparams systemStart eraHistory stakePools
   forM txs $ \tx -> do
     submitTransaction networkId socketPath tx
     void $ awaitTransaction networkId socketPath tx
     pure $ txId tx
  where
   vk = getVerificationKey sk
+  buildHydraScriptTxs pparams systemStart eraHistory stakePools = do
+    utxo <- queryUTxOFor networkId socketPath QueryTip vk
+    buildScriptPublishingTxs pparams systemStart networkId eraHistory stakePools utxo sk
+      `catch` \(ex :: PublishScriptException) ->
+        case ex of
+          FailedToBuildPublishingTx _ -> throwIO ex
+          FailedToSquashUTxOToCoverDeposit _ -> throwIO ex
+          FailedToFindUTxOToCoverDeposit totalDeposit -> do
+            let totalUTxOValue = selectLovelace $ foldMap (txOutValue . snd) (UTxO.pairs utxo)
+            if totalUTxOValue < totalDeposit
+              then throwIO ex
+              else do
+                -- XXX: squash the utxo and retry
+                rawSquashUTxOTx <- buildSquashUTxOTx pparams systemStart networkId eraHistory stakePools vk utxo totalDeposit
+                let squashUTxOTx = signTx sk rawSquashUTxOTx
+                submitTransaction networkId socketPath squashUTxOTx
+                void $ awaitTransaction networkId socketPath squashUTxOTx
+                buildHydraScriptTxs pparams systemStart eraHistory stakePools
```

Comment:
This will be an endless loop!

We should be doing ``build `catch` squash >> build`` only once